src/Compilers/Core/Portable/Diagnostic/DiagnosticSeverity.cs

```diff
@@ -27,7 +27,12 @@ public enum DiagnosticSeverity
         /// <summary>
         /// Something not allowed by the rules of the language or other authority.
         /// </summary>
-        Error = 3
+        Error = 3,
+
+        /// <summary>
+        /// Something that can be potentially improved.
+        /// </summary>
+        Suggestion = 4,
```

Comment:
>Well, mistakes in past doesn't justify repeating them, especially if we know we could have done it better :)

True. But, if the past mistakes are so prevalent in the existing design, those mistakes can actually _become_ the design. And, trying to do improve upon that results in inconsistency. I believe that's the case here.

> If "Suggestion" gets the vote for being a new severity, why not "Unnecessary"?

I keep repeating this, but "Unnecessary" is _intended_ to be combined with a severity. It is 100% by design that code could be faded as "unnecessary" and be squiggled as an error. This was designed from the very beginning. Likewise, code can be unnecessary and marked as a suggestion.

>I don't agree with this. Firstly, custom tags are associated with descriptors, i.e. rule metadata, not individual diagnostic instances. Rule metadata represents default information about category of issues created through the descriptor, as deemed appropriate by the analyzer author. Even the unnecessary custom tag added to faded diagnostic becomes non-sensical if user promotes the diagnostic to an error. We need to understand that custom tags is just metadata for class of diagnostics reported via a descriptor.

How is it non-sensical if a user promotes a faded diagnostic to an error? If the user wants to promote remove redundant cast to an error, they should be able to do that and the diagnostic should continue to be faded (because it represents unnecessary code) and underlined with a red squiggle (because it is an error).

That said, "suggestion" is a different concept that "unnecessary". "Errors", "Warnings", and "Suggestions" are mutually exclusive, which implies that using a custom tag for "Suggestion" is a fragile design with invalid combinations.

To be super clear, we're not suggesting that we should allow the DiagnosticSeverity enum to explode. However, after spending the last couple of years building analyzers, and looking at the problem space we believe that there is simply a missing severity. This severity is not so strong as to warrant a warning to inform the user that something is truly wrong with their code. Instead, it is slightly weaker, offering a suggestion to improve the user's code.

You're completely right that suggestions could be achieved by using the Info severity with a custom tag. However, I would argue that we'd just be simulating a severity in that case. And, if we're contemplating the simulation of a severity, I argue that it deserves to be a severity. Otherwise, it becomes an extra complication that every diagnostic host needs handle properly.